framework/src/main/java/org/checkerframework/framework/util/ComplexQualifierHierarchy.java

```diff
@@ -0,0 +1,375 @@
+        if (lub.hasElements()) {
+            return leastUpperBound(a1, qual1, a2, qual2);
+        }
+        return qualifierMap.get(lub);
+    }
+
+    /**
+     * Returns the least upper bound between {@code a1} and {@code a2}.
+     *
+     * <p>This method is only called when the lub is an annotation with elements.
+     *
+     * @param a1 first annotation
+     * @param qual1 QualifierKind for {@code a1}
+     * @param a2 second annotation
+     * @param qual2 QualifierKind for {@code a2}
+     * @return the least upper bound between {@code a1} and {@code a2}.
+     */
+    protected abstract AnnotationMirror leastUpperBound(
+            AnnotationMirror a1, QualifierKind qual1, AnnotationMirror a2, QualifierKind qual2);
+
+    @Override
+    public AnnotationMirror greatestLowerBound(AnnotationMirror a1, AnnotationMirror a2) {
+        QualifierKind qual1 = getQualifierKind(a1);
+        QualifierKind qual2 = getQualifierKind(a2);
+        if (!qual1.areInSameHierarchy(qual2)) {
+            return null;
+        }
+        QualifierKind glb = qualifierKindHierarchy.greatestLowerBound(qual1, qual2);
+        if (glb.hasElements()) {
+            return greatestLowerBound(a1, qual1, a2, qual2);
+        }
+        return qualifierMap.get(glb);
+    }
+
+    /**
+     * Returns the greatest lower bound between {@code a1} and {@code a2}.
+     *
+     * <p>This method is only called when the glb is an annotation with elements.
+     *
+     * @param a1 first annotation
+     * @param qual1 QualifierKind for {@code a1}
+     * @param a2 second annotation
+     * @param qual2 QualifierKind for {@code a2}
+     * @return the greatest lower bound between {@code a1} and {@code a2}.
+     */
+    protected abstract AnnotationMirror greatestLowerBound(
+            AnnotationMirror a1, QualifierKind qual1, AnnotationMirror a2, QualifierKind qual2);
+
+    @Override
+    public boolean isSubtypeTypeVariable(AnnotationMirror subAnno, AnnotationMirror superAnno) {
+        if (superAnno == null) {
+            // [] is a supertype of any qualifier, and [] <: []
+            return true;
+        }
+        if (subAnno == null) {
+            // [] is a subtype of no qualifier (only [])
+            return false;
+        }
+        return isSubtype(subAnno, superAnno);
+    }
+
+    @Override
+    public boolean isSubtypeTypeVariable(
+            Collection<? extends AnnotationMirror> subAnnos,
+            Collection<? extends AnnotationMirror> superAnnos) {
+        int isSubtypeCount = 0;
+        for (AnnotationMirror subAnno : subAnnos) {
+            QualifierKind subKind = getQualifierKind(subAnno);
+            for (AnnotationMirror superAnno : superAnnos) {
+                QualifierKind superKind = getQualifierKind(superAnno);
+                if (subKind.areInSameHierarchy(superKind)) {
+                    if (isSubtype(subAnno, superAnno)) {
+                        isSubtypeCount++;
```

Comment:
It is more efficient to `break;` right after this increment operation.

(Alternately, you could abstract out the inner loop into a separate method, and there would be no need for a count variable.  It might be a bit less efficient because of the method calls, but would be more efficient in being able to exit early rather than checking all type hierarchies.)

Same comment applies in 4 other places in the pull request.